relations/src/linear.rs

```diff
@@ -0,0 +1,54 @@
+use ark_ff::PrimeField;
+use ark_r1cs_std::{
+    prelude::{AllocVar, EqGadget},
+    uint32::UInt32,
+};
+use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError};
+
+use crate::{relation::GetPublicInput, CircuitField};
+
+/// Linear equation relation: a*x + b = y
+///
+/// Relation with:
+///  - 1 private witness (x)
+///  - 3 constants        (a, b, y)
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct LinearEquationRelation {
+    /// constant (a slope)
+    pub a: u32,
+    /// private witness
+    pub x: u32,
+    /// constant(an intercept)
+    pub b: u32,
+    /// constant
+    pub y: u32,
+}
+
+impl LinearEquationRelation {
+    pub fn new(a: u32, x: u32, b: u32, y: u32) -> Self {
+        Self { a, x, b, y }
+    }
+}
+
+impl<Field: PrimeField> ConstraintSynthesizer<Field> for LinearEquationRelation {
+    fn generate_constraints(self, cs: ConstraintSystemRef<Field>) -> Result<(), SynthesisError> {
+        // Watch out for overflows!!!
+        let x = UInt32::new_witness(ark_relations::ns!(cs, "x"), || Ok(&self.x))?;
```

Comment:
please add todo for migrating from `UInt32` to `FpVar` (it will also make `impl GetPublicInput<CircuitField> for LinearEquationRelation` easier to implement)

Reply:
the same for `xor` as well

Reply:
:ok_hand: